src/idl_gen_fbs.cpp

```diff
@@ -39,6 +43,97 @@ static std::string GenType(const Type &type, bool underlying = false) {
   }
 }
 
+static std::unordered_map<std::string, voffset_t> MapStructId(
+    const StructDef &struct_def, IDLOptions::ProtoIdGapAction gap_action) {
+  using FieldIdNamePair = std::pair<std::string, std::string>;
+  std::vector<FieldIdNamePair> ids;
+
+  bool field_with_id = false;
+
+  for (const auto &field : struct_def.fields.vec) {
+    if (field->attributes.Lookup("id")) {
+      field_with_id = true;
+      if (field->attributes.Lookup("id")->constant.empty() &&
+          field->value.type.base_type != BASE_TYPE_UNION)
+        ids.push_back(std::make_pair("-1", field->name));
+      else {
+        ids.push_back(std::make_pair(field->attributes.Lookup("id")->constant,
+                                     field->name));
+
+        // Check for non positive id number
+        if (!field->attributes.Lookup("id")->constant.empty()) {
+          voffset_t id = 0;
+          bool done = StringToNumber(
+              field->attributes.Lookup("id")->constant.c_str(), &id);
+          if (!done)
+            throw std::runtime_error("Field id in struct" + struct_def.name +
+                                     "has a non positive number value\n");
+        }
+      }
+    } else {
+      ids.push_back(std::make_pair(std::string(), field->name));
```

Comment:
Not immediately sure what pushing these empty strings is doing. Should we ever have a case where the field ID is unset? And if we do, is there any reason to store it and not just skip over it?

Reply:
Unions do not have id. We can not skip them